backend/base/mdns/detector.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/sirupsen/logrus"
+)
+
+// Detector listens for new bitboxbases and creates a new bitboxBase when detected.
+type Detector struct {
+	baseDeviceInterface map[string]bitboxbase.Interface
+
+	onRegister   func(bitboxbase.Interface) error
+	onUnregister func(string)
+
+	removedIDs map[string]bool //keeps track of the bases that were removed manually by the user.
+
+	log *logrus.Entry
+}
+
+// NewDetector creates a new detector. onRegister is called when a bitboxbase has been
+// inserted. onUnregister is called when the bitboxbase has been removed.
+//
+func NewDetector(
+	onRegister func(bitboxbase.Interface) error,
+	onUnregister func(string),
+) *Detector {
+	return &Detector{
+		baseDeviceInterface: map[string]bitboxbase.Interface{},
+		onRegister:          onRegister,
+		onUnregister:        onUnregister,
+		removedIDs:          make(map[string]bool),
+
+		log: logging.Get().WithGroup("detector"),
+	}
+}
+
+// TryMakeNewBase attempts to create a new bitboxBase connection to the BitBox base. Returns true if successful, false otherwise.
+func (detector *Detector) TryMakeNewBase(ip string) (bool, error) {
+	for bitboxBaseID, bitboxBase := range detector.baseDeviceInterface {
+		// Check if bitboxbase was removed.
+		if detector.checkIfRemoved(bitboxBaseID) {
+			bitboxBase.Close()
+			delete(detector.baseDeviceInterface, bitboxBaseID)
+			detector.onUnregister(bitboxBaseID)
+			detector.log.WithField("bitboxbase-id", bitboxBaseID).Info("Unregistered bitboxbase")
+		}
+	}
+
+	// Make the id the ip for now, later the pairing should be factored in here as well.
+	bitboxBaseID := ip
+	// Skip if already registered.
+	if _, ok := detector.baseDeviceInterface[bitboxBaseID]; ok {
+		return false, nil
+	}
+	var baseDevice bitboxbase.Interface
+	var err error
+	baseDevice, err = base.NewBitBoxBase(ip, bitboxBaseID)
+
+	if err != nil {
+		detector.log.WithError(err).Error("Failed to register Base")
+		return false, err
+	}
+	detector.baseDeviceInterface[bitboxBaseID] = baseDevice
+	if err := detector.onRegister(baseDevice); err != nil {
+		detector.log.WithError(err).Error("Failed to execute on-register")
+	}
+	return true, nil
+}
+
+// RemoveBase allows external objects to delete a detector entry.
+func (detector *Detector) RemoveBase(bitboxBaseID string) {
+	if _, ok := detector.baseDeviceInterface[bitboxBaseID]; ok {
+		delete(detector.baseDeviceInterface, bitboxBaseID)
+		detector.removedIDs[bitboxBaseID] = true
+	}
+}
+
+// checkIfRemoved returns true if a bitboxbase was detected in, but is not reachable anymore.
+func (detector *Detector) checkIfRemoved(bitboxBaseID string) bool {
+	//TODO implement this function, should check if the bitboxBase is reachable, if not properly unregister it.
+	return false
+}
+
+func (detector *Detector) listen() {
+	for {
+		//TODO This is where the mdns implementation should hook in, code is kept here to give an idea of the interfaces involved.
+
+		resolver, err := zeroconf.NewResolver(nil)
```

Comment:
move instance of this to the struct instead of creating a new one every iteration?

Reply:
It can't seem to resolve mdns connections anymore, when I create the resolver outside of the loop. I'm a bit puzzled why that is the case.

Reply:
I changed the code a bit here, there is now a separate function that starts a goroutine with a new context. The goroutine returns once <- context.Done() is triggered.